pkg/sql/schemachanger/scbuild/build.go

```diff
@@ -441,3 +471,16 @@ func (b buildCtx) WithNewSourceElementID() scbuildstmt.BuildCtx {
 		EventLogState: b.EventLogStateWithNewSourceElementID(),
 	}
 }
+
+// shouldElementBeRetainedWithoutMetadata tracks which elements should
+// be retained even if no metadata exists. These elements may contain
+// other hints that are used for planning such as TableData/IndexData elements
+// that allow us to skip the two version invariant or backfills/validation
+// at runtime.
+func shouldElementBeRetainedWithoutMetadata(element scpb.Element) bool {
+	switch element.(type) {
+	case *scpb.TableData, *scpb.IndexData:
+		return true
+	}
+	return false
+}
```

Comment:
I wonder if perhaps we'd be better served by adding a boolean override flag in the metadata protobuf instead, that would be set by `scdecomp` for TableData/IndexData elements. I'm not sure if it's actually better, perhaps not. Your approach is fine in the meantime, perhaps I'll explore this after this merges.